### ISSUE TITLE
Asyncio TimeoutError in AD4

### ISSUE THREAD

**Issue by bachya** [AppDaemon/appdaemon#852]
(perhaps related to https://github.com/home-assistant/appdaemon/issues/848)

I have an app that calls a service at random intervals between 5 and 60 minutes:

```python
from random import randint

from appdaemon.plugins.hass.hassapi import Hass


class ServiceOnRandomTick(Hass):  # pylint: disable=too-few-public-methods
    """Define an automation to call a service at random moments."""

    def initialize(self) -> None:
        """Initialize."""
        self.run_every(
            self._on_tick, self.datetime(), randint(5 * 60, 60 * 60),
        )

    def _on_tick(self, kwargs: dict) -> None:
        """Fire the event when the tick occurs."""
        self.call_service(self.args["service"], **self.args["service_data"])
```

It is configured like this:

```yaml
living_room_lights_randomize_automation:
  module: services
  class: ServiceOnRandomTick
  service: "switch/toggle"
  service_data:
    entity_id: group.living_room_lights
```

When the service call executes, the intended action – in the case of the above, toggling `group.living_room_lights` – works as expected in AD4; however, I also see this exception in the logs:

```
2020-01-19 20:35:02.016359 WARNING living_room_lights_randomize_automation: ------------------------------------------------------------
2020-01-19 20:35:02.016766 WARNING living_room_lights_randomize_automation: Unexpected error in worker for App living_room_lights_randomize_automation:
2020-01-19 20:35:02.017104 WARNING living_room_lights_randomize_automation: Worker Ags: {'id': 'ef57c0d255444c1cb5d727281ac4c846', 'name': 'living_room_lights_randomize_automation', 'objectid': 'cb0c8c636fa140a0844bb8f60c63d756', 'type': 'scheduler', 'function': <bound method ServiceOnRandomTick._on_tick of <services.ServiceOnRandomTick object at 0x7fc4fdaf9b80>>, 'pin_app': True, 'pin_thread': 18, 'kwargs': {'interval': 2872, '__thread_id': 'thread-18'}}
2020-01-19 20:35:02.017402 WARNING living_room_lights_randomize_automation: ------------------------------------------------------------
2020-01-19 20:35:02.019121 WARNING living_room_lights_randomize_automation: Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/appdaemon/threading.py", line 766, in worker
    funcref(self.AD.sched.sanitize_timer_kwargs(app, args["kwargs"]))
  File "/conf/apps/services.py", line 76, in pick_and_call_service
    self.call_service(self.args["service"], **self.args["service_data"])
  File "/usr/local/lib/python3.8/site-packages/appdaemon/utils.py", line 191, in inner_sync_wrapper
    f = run_coroutine_threadsafe(self, coro(self, *args, **kwargs))
  File "/usr/local/lib/python3.8/site-packages/appdaemon/utils.py", line 285, in run_coroutine_threadsafe
    result = future.result(self.AD.internal_function_timeout)
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 441, in result
    raise TimeoutError()
concurrent.futures._base.TimeoutError

2020-01-19 20:35:02.019420 WARNING living_room_lights_randomize_automation: ------------------------------------------------------------
2020-01-19 20:35:02.033111 WARNING AppDaemon: callback _on_tick() in living_room_lights_randomize_automation has now completed
```

**Comment by Odianosen25** [comment]
Hello @bachya,

It seems to me, the service call to HA is creating a significant delay when executing that call sometimes. It’s actually different from the other issue. When a service call is made, a response is expected, and since it’s a group, HA might be taking time to run. This could hook up the request. 

The fact the last line references that it has now completed, means there was a delay somewhere in your apps causing this issue. So for that reason, the task was cancelled.

If you use async apps, this might be a result of another async app holding on the loop. Also if you use register_service, one of your service calls could have held up AD’s call service, and prevented this one from running smoothly. But I doubt is any of the above reasons.

My cash is on the fact connection between AD and HA are sometimes not best of friends, and to avoid AD cores from being held, it cancels the function after 10 seconds. 

There is internal discussion to possibly move service calls to HA using websockets instead of requests to avoid things like this, or to drop the idea of requesting a result when a service call is made. But in all, kindly check if your HA is not too busy, and it’s holding up processing that request quickly. 

It might even be an issue between HA and the lights in question, and the delay is passed back to AD who then freaks out.

Hope this helps. Regards

**Comment by bachya** [comment]
@Odianosen25 Thanks for your reply. Some more info that leads me to believe this is a bug (and not something environmental):

* I am not currently using any async apps.
* this exact app did not throw an exception in AD 3.x.
* I don’t think “HASS busyness” is the problem. When I query this same service call over the REST API (using cURL), both the group toggling and the response payload are immediate.

**Comment by acockburn** [comment]
This looks like some cruft from our changes to allow async apps - we'll take a look, but from what I understand at the moment this is just cosmetic?

**Comment by bachya** [comment]
@acockburn So far as I can tell. 👍🏻

**Comment by bg1000** [comment]
@acockburn I am seeing this issue as well. I am able to work around it but I think it's more than cosmetic.  I have an automation that turns off my (simplisafe) alarm system when I arrive home like so:
`                self.call_service("alarm_control_panel/alarm_disarm", entity_id="alarm_control_panel.XXX")`

The HASS Simplisafe integration uses the (undocumented) Simplisafe cloud api and it does respond rather slowly. Under Appdaemon 3.05 this would work 95-99% of the time.  Under Appdaemon 4.0.0 I see the TimeoutError just about every single time.  I can do some further testing to confirm if it would be helpful but I do not think the remaining code in the callback runs after the error occurs. 

I would agree that the response time from Simplisafe is too long and there are workarounds available but it would be helpful, in my opinion, if Appdaemon dealt with this a little differently.
I tried this (maybe I didn't catch the right exception or had a syntax error I didn't catch?) which did not seem to work:
```
try:
    self.call_service . . . 
except TimeoutError:
    self.log . . . 
```

As it stands currently I am setting an input_boolean from AppDaemon which triggers a HASS automation. 

Bottom Line: This might not be critical but I would appreciate it if you could keep it on the list of things to look at for the next release (or the one after that).

**Comment by ReneTode** [comment]
with adding asyncio to AD we have added an issue that services that hold up longer, also hold up ADs core.
this is an issue we know and try to find and solve as soon as possible, because its a very annoying issue that can effect everyone.

thx. for explaining it so good @bg1000 because untill now i thought it only affected the new options to create services internal in AD. this makes clear that the service holdup problem is in all services.

**Comment by bg1000** [comment]
@ReneTode, Thanks for replying.  I run in docker and would be happy to pull a dev container for testing and/or verification.  If the team gets to a point where that would be helpful feel free to hit me up here or on the home assistant community forum (same username there).

**Comment by so3n** [comment]
I think I'm also having the same issue, which I only just noticed since upgrading to AD4. I don't have any async apps.

I'm able to recreate the issue with this simple app that turns on my tv using harmony hub component. HA usually takes a while to register that harmony hub state is on (maybe a bit longer than 10 sec) but this has always been the case, even with AD3 (where I didn't experience this issue)

```py
import appdaemon.plugins.hass.hassapi as hass

class test(hass.Hass):
    def initialize(self):
        self.turn_on("remote.living_room")
```

```bash
2020-02-04 12:05:54.105218 INFO AppDaemon: Initializing app test123 using class test from module other
2020-02-04 12:06:04.166921 WARNING test123: ------------------------------------------------------------
2020-02-04 12:06:04.169795 WARNING test123: Unexpected error running initialize() for test123
2020-02-04 12:06:04.172966 WARNING test123: ------------------------------------------------------------
2020-02-04 12:06:04.181620 WARNING test123: Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/appdaemon/app_management.py", line 145, in initialize_app
    await utils.run_in_executor(self, init)
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 276, in run_in_executor
    response = future.result()
asyncio.exceptions.TimeoutError

2020-02-04 12:06:04.182943 WARNING test123: ------------------------------------------------------------
2020-02-04 12:06:04.184698 WARNING AppDaemon: Logged an error to /config/appdaemon/logs/appdaemon.log
2020-02-04 12:06:04.196062 WARNING AppDaemon: Excessive time spent in utility loop: 10698.0ms, 10694.0ms in check_app_updates(), 4.0ms in other
```
```

**Comment by Odianosen25** [comment]
@so3n, 

Thanks for this, and this confirms an earlier suspicion of the fact that a delayed service call execution within HA, will produce the error also.

We will look into a fix for this.

Regards

**Comment by gibman** [comment]
also seeing this issue on code that previously ran rock steady on AD3.

It happened this morning when I turned on a harmony remote "activity".


error.log:

`2020-03-03 08:11:16.490938 WARNING harmony: ------------------------------------------------------------
2020-03-03 08:11:16.493788 WARNING harmony: Unexpected error in worker for App harmony:
2020-03-03 08:11:16.496610 WARNING harmony: Worker Ags: {'id': 'f99acaf640ae4a6498a118ca7ea8ea3c', 'name': 'harmony', 'objectid': '0e31175b2c3c4212976b8d2d224a9615', 'type': 'state', 'function': <bound method harmony.harmony_on_activity_changed of <harmony.harmony object at 0x731367d8>>, 'attribute': 'current_activity', 'entity': 'remote.livingroom', 'new_state': '(TV) Netflix', 'old_state': 'PowerOff', 'pin_app': True, 'pin_thread': 20, 'kwargs': {'attribute': 'current_activity', '__thread_id': 'thread-20'}}
2020-03-03 08:11:16.498965 WARNING harmony: ------------------------------------------------------------
2020-03-03 08:11:16.503031 WARNING harmony: Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/appdaemon/threading.py", line 777, in worker
    funcref(entity, attr, old_state, new_state,
  File "/config/appdaemon/apps/harmony.py", line 1382, in harmony_on_activity_changed
    self.on_htpc_comleted_harmony_activity()
  File "/config/appdaemon/apps/harmony.py", line 1515, in on_htpc_comleted_harmony_activity
    self.turn_on_htpc()
  File "/config/appdaemon/apps/harmony.py", line 813, in turn_on_htpc
    self.call_service("switch/turn_on", entity_id = "switch.livingroom_htpc_wol")
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 191, in inner_sync_wrapper
    f = run_coroutine_threadsafe(self, coro(self, *args, **kwargs))
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 285, in run_coroutine_threadsafe
    result = future.result(self.AD.internal_function_timeout)
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 441, in result
    raise TimeoutError()
concurrent.futures._base.TimeoutError`


harmony event listener:
`self.listen_state(self.harmony_on_activity_changed, self._harmony, attribute=self.CURRENT_ACTIVITY)
`


culprit method:
`def turn_on_htpc(self, kwargs = None):            
  self.call_service("switch/turn_on", entity_id = "switch.livingroom_htpc_wol")
`

configuration.yaml:
`  
  - platform: wake_on_lan
    mac: "4C-CC-6A-FD-EE-4D"
    name: livingroom htpc WOL
    turn_off: 
      service: mqtt.publish 
      data_template:
        topic: "enkegaard/firstfloor/htpc/on_before_poweroff"    
        payload: "suspend"  `


so the WOL uses a local mosquito MQTT server instance.

**Comment by acockburn** [comment]
I’m still of the opinion that this is just exposing previously long running calls that we are now trapping and error if instead of waiting indefinitely - I think the thing to do here at a minimum is provide a timeout parameter so it can be set higher or even to forever to restore the previous v3 behavior

**Comment by jacekpaszkowski** [comment]
I often get timeout when calling HA service or changing state...

 ```
020-03-05 19:06:30.392156 WARNING salon_purifier_controller: Unexpected error in worker for App salon_purifier_controller:
2020-03-05 19:06:30.395367 WARNING salon_purifier_controller: Worker Ags: {'id': '888bec6b2db94c9c947d25720e867ed5', 'name': 'salon_purifier_controller', 'objectid': '4154f0922e1f4a688b0400fe9b1d4c67', 'type': 'state', 'function': <bound method SalonPurifierController.fan_state_changed of <salon_purifier_controller.SalonPurifierController object at 0x75347f40>>, 'attribute': 'state', 'entity': 'fan.salon_air_purifier', 'new_state': 'on', 'old_state': 'off', 'pin_app': True, 'pin_thread': 9, 'kwargs': {'new': 'on', '__thread_id': 'thread-9'}}
2020-03-05 19:06:30.397263 WARNING salon_purifier_controller: ------------------------------------------------------------
2020-03-05 19:06:30.401616 WARNING salon_purifier_controller: Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/appdaemon/threading.py", line 777, in worker
    funcref(entity, attr, old_state, new_state,
  File "/config/appdaemon/apps/salon_purifier_controller.py", line 131, in fan_state_changed
    self.reset_variables()
  File "/config/appdaemon/apps/salon_purifier_controller.py", line 139, in reset_variables
    self.call_service('var/set', entity_id = self.var_manual_on, value = 'off')
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 191, in inner_sync_wrapper
    f = run_coroutine_threadsafe(self, coro(self, *args, **kwargs))
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 285, in run_coroutine_threadsafe
    result = future.result(self.AD.internal_function_timeout)
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 441, in result
    raise TimeoutError()
concurrent.futures._base.TimeoutError
```

**Comment by jacekpaszkowski** [comment]
Alos sometimes when reloading app

```
2020-03-05 21:14:46.252455 INFO AppDaemon: Terminating salon_purifier_controller
2020-03-05 21:14:46.263912 INFO AppDaemon: Reloading Module: /config/appdaemon/apps/salon_purifier_controller.py
2020-03-05 21:14:46.337859 INFO AppDaemon: Initializing app salon_purifier_controller using class SalonPurifierController from module salon_purifier_controller
2020-03-05 21:14:56.372105 WARNING salon_purifier_controller: ------------------------------------------------------------
2020-03-05 21:14:56.374471 WARNING salon_purifier_controller: Unexpected error running initialize() for salon_purifier_controller
2020-03-05 21:14:56.377693 WARNING salon_purifier_controller: ------------------------------------------------------------
2020-03-05 21:14:56.387701 WARNING salon_purifier_controller: Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/appdaemon/app_management.py", line 145, in initialize_app
    await utils.run_in_executor(self, init)
  File "/usr/lib/python3.8/site-packages/appdaemon/utils.py", line 276, in run_in_executor
    response = future.result()
asyncio.exceptions.TimeoutError
``

**Comment by so3n** [comment]
Any fix in sight for this? Or is the best option to just go back to AD3?

**Comment by jason0x43** [comment]
You can set the `internal_function_timeout` property to a higher value to get around this. In `appdaemon.yaml`:

```
appdaemon:
  internal_function_timeout: 30
```

The value is in seconds. It defaults to 10.

**Comment by jason0x43** [comment]
There's also `thread_duration_warning_threshold` if you don't like seeing a warning every time a long-running task completes:

```
appdaemon:
  thread_duration_warning_threshold: 20
```

Value is seconds, default is 10.

**Comment by so3n** [comment]
Thanks thats fixed the issue for me

**Comment by Odianosen25** [comment]
@so3n, as much as I am glad that solved your issue, sometimes increasing the `internal_function_timeout` from the default 10 seconds is not the best option. There might be an underlining serious issue, which increasing the timeout only masks it.

I have had instances where I had the timeout error, and I increased mine to `60` and though "solved" my error log having the error, but digging more, found out it was when updating an entity in HA for some reasons was not being updated. I found this out, because the same python module was used by another app and that gave me no issue. I simply manually reset the entity in HA's developer tools and it disappeared. 

So in all, might be worthwhile to dig thru, and optimise your code, or at least try and see what the issue is. I am not saying there are no legitimate reasons to increase it, especially if running some tasks that cuts across an app's thread to the core and back. If that is the case, a re-design of the system should be considered, than increasing the timeout.

Regards
